### PR TITLE
[SPARK-14369][SQL] Locality support for FileScanRDD

### PR DESCRIPTION
**Pull request by liancheng**
## What changes were proposed in this pull request?

This PR adds preliminary locality support for `FileFormat` data sources by overriding `FileScanRDD.preferredLocations()`. The strategy can be divided into two parts:
1.  Block location lookup
   
   Unlike `HadoopRDD` or `NewHadoopRDD`, `FileScanRDD` doesn't have access to the underlying `InputFormat` or `InputSplit`, and thus can't rely on `InputSplit.getLocations()` to gather locality information. Instead, this PR queries block locations using `FileSystem.getBlockLocations()` after listing all `FileStatus`es in `HDFSFileCatalog` and convert all `FileStatus`es into `LocatedFileStatus`es.
   
   Note that although S3/S3A/S3N file systems don't provide valid locality information, their `getLocatedStatus()` implementations don't actually issue remote calls either. So there's no need to special case these file systems.
2.  Selecting preferred locations
   
   For each `FilePartition`, we pick up top 3 locations that containing the most data to be retrieved. This isn't necessarily the best algorithm out there. Further improvements may be brought up in follow-up PRs.
## How was this patch tested?

Tested by overriding default `FileSystem` implementation for `file:///` with a mocked one, which returns mocked block locations.

### CODE REVIEW

**Comment by JoshRosen** [comment]
FYI, I don't think you can necessarily put multiple `test-hadoop` commands in the same PR title and have it run multiple tests.

**Comment by liancheng** [comment]
@JoshRosen Oh, thanks for reminding. Then I'll test different Hadoop versions manually by changing PR titles.

**Comment by liancheng** [comment]
retest this please

**Comment by liancheng** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala [hunk 1]
```diff
@@ -620,17 +621,28 @@ class HDFSFileCatalog(
...
         val pathFilter = FileInputFormat.getInputPathFilter(jobConf)
-        if (pathFilter != null) {
+        val statuses = if (pathFilter != null) {
           Try(fs.listStatus(path, pathFilter)).getOrElse(Array.empty)
         } else {
           Try(fs.listStatus(path)).getOrElse(Array.empty)
         }
+
+        fs match {
+          case _: S3FileSystem =>
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala]: S3AFileSystem / NativeS3FileSystem. S3FileSystem is for the block store.

File: sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala [hunk 2]
```diff
@@ -620,17 +621,28 @@ class HDFSFileCatalog(
...
+            // S3 doesn't provide locality information.
+            statuses.toSeq
+
+          case _ =>
+            statuses.map { f =>
+              new LocatedFileStatus(f, fs.getFileBlockLocations(f, 0, f.getLen))
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala]: At here, let's check if this file status is already a LocatedFileStatus.

File: sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala (hunk 1 above)
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala]: Actually, do we need to have this pattern matching? Looks like S3 filesystems just use the default implementation of getFileBlockLocations, which returns `localhost:50010`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileScanRDD.scala [hunk 3]
```diff
@@ -86,4 +87,8 @@ class FileScanRDD(
   }
 
   override protected def getPartitions: Array[Partition] = filePartitions.toArray
+
+  override protected def getPreferredLocations(split: Partition): Seq[String] = {
+    split.asInstanceOf[FilePartition].files.flatMap(_.locations).distinct
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileScanRDD.scala]: Looks we want to pick the location based on the size of a block? In MapOutputTracker we have `getPreferredLocationsForShuffle`, which can be related to us.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala [hunk 4]
```diff
@@ -137,10 +140,12 @@ private[sql] object FileSourceStrategy extends Strategy with Logging {
 
           val splitFiles = selectedPartitions.flatMap { partition =>
             partition.files.flatMap { file =>
+              val blockLocations = getBlockLocations(file)
               (0L to file.getLen by maxSplitBytes).map { offset =>
                 val remaining = file.getLen - offset
                 val size = if (remaining > maxSplitBytes) maxSplitBytes else remaining
-                PartitionedFile(partition.values, file.getPath.toUri.toString, offset, size)
+                val hosts = getBlockHosts(blockLocations, offset)
+                PartitionedFile(partition.values, file.getPath.toUri.toString, offset, size, hosts)
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala]: How about we add document to this FileSourceStrategy to explain our strategy? So, readers can quickly understand what we are doing at here.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala [hunk 5]
```diff
@@ -205,4 +210,20 @@ private[sql] object FileSourceStrategy extends Strategy with Logging {
 
     case _ => Nil
   }
+
+  private def getBlockLocations(file: FileStatus): Array[BlockLocation] = file match {
+    case f: LocatedFileStatus => f.getBlockLocations
+    case f => Array.empty[BlockLocation]
+  }
+
+  private def getBlockIndex(blockLocations: Array[BlockLocation], offset: Long): Int = {
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala]: What is the semantic of `offset`?

File: sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala (hunk 2 above)
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala]: Done.

File: sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala (hunk 1 above)
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala]: Thanks, made corresponding changes and added comments.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileScanRDD.scala (hunk 3 above)
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileScanRDD.scala]: Added a constraint so that we only pick locations of blocks whose sizes are larger than 10% of the whole file partition.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala (hunk 4 above)
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala]: Added.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala (hunk 5 above)
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala]: Added comments.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala [hunk 6]
```diff
@@ -137,10 +140,14 @@ private[sql] object FileSourceStrategy extends Strategy with Logging {
...
             partition.files.flatMap { file =>
+              val blockLocations = getBlockLocations(file)
               (0L to file.getLen by maxSplitBytes).map { offset =>
                 val remaining = file.getLen - offset
                 val size = if (remaining > maxSplitBytes) maxSplitBytes else remaining
-                PartitionedFile(partition.values, file.getPath.toUri.toString, offset, size)
+                // Finds out a list of location hosts where lives the first block that contains the
+                // starting point the file block starts at `offset` with length `size`.
+                val hosts = getBlockHosts(blockLocations, offset)
+                PartitionedFile(partition.values, file.getPath.toUri.toString, offset, size, hosts)
```
Comment (nongli) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala]: How is this locality information maintained when we coalesce in the logic below?

Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala]: Each `FilePartition` derives their own preferred location in `FileScanRDD.preferredLocations()` from location hosts of all its `PartitionedFile`s.

Comment (nongli) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala]: This is not likely to generate good locality right?

Is it reasonable to repurpose this?
https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/rdd/CoalescedRDD.scala#L124

Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala]: This is a really good pointer, thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileScanRDD.scala [hunk 7]
```diff
@@ -86,4 +87,20 @@ class FileScanRDD(
   }
 
   override protected def getPartitions: Array[Partition] = filePartitions.toArray
+
+  override protected def getPreferredLocations(split: Partition): Seq[String] = {
+    val files = split.asInstanceOf[FilePartition].files
+    val partitionSize = files.map(_.length).sum
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileScanRDD.scala]: how about:
```
val hostToLength = Map[String, Long]()
for (file <- files) {
  for (host <- file.hosts) {
    if (hostToLength.contains(host)) {
      hostToLength(host) += file.length
    } else {
      hostToLength(host) = file.length
    }
  }
  hostToLength.toSeq.sortBy(_._2).take(3).map(_._1)
}
```

File: sql/hive/src/test/scala/org/apache/spark/sql/sources/hadoopFsRelationSuites.scala [hunk 8]
```diff
@@ -668,6 +671,42 @@ abstract class HadoopFsRelationTest extends QueryTest with SQLTestUtils with Tes
...
+        val df1 = sqlContext.range(4)
+        df1.coalesce(1).write.mode("overwrite").format(dataSourceName).save(path)
+        df1.coalesce(1).write.mode("append").format(dataSourceName).save(path)
+
+        val df2 = sqlContext.read
+          .format(dataSourceName)
+          .option("dataSchema", df1.schema.json)
+          .load(path)
+
+        val Some(fileScanRDD) = {
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/sources/hadoopFsRelationSuites.scala]: how about:
```
val fileScanRDD = df2.queryExecution.executedPlan.collectFirst {
  case scan: DataSourceScan if scan.rdd.isInstanceOf[FileScanRDD] => scan.rdd.asInstanceOf[FileScanRDD]
}
```

File: sql/hive/src/test/scala/org/apache/spark/sql/sources/hadoopFsRelationSuites.scala [hunk 9]
```diff
@@ -668,6 +671,42 @@ abstract class HadoopFsRelationTest extends QueryTest with SQLTestUtils with Tes
...
+          }
+
+          // We have to search for the `FileScanRDD` along the RDD DAG since
+          // RDD.preferredLocations doesn't propagate along the DAG to the root RDD.
+          allRDDsInDAG(df2.rdd).collectFirst {
+            case f: FileScanRDD => f
+          }
+        }
+
+        val partitions = fileScanRDD.partitions
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/sources/hadoopFsRelationSuites.scala]: do we always return 2 partitions?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala [hunk 10]
```diff
@@ -197,4 +204,26 @@ private[sql] object FileSourceStrategy extends Strategy with Logging {
...
+
+  private def getBlockLocations(file: FileStatus): Array[BlockLocation] = file match {
+    case f: LocatedFileStatus => f.getBlockLocations
+    case f => Array.empty[BlockLocation]
+  }
+
+  // Given locations of all blocks of a single file, `blockLocations`, and an `offset` position of
+  // the same file, find out the index of the block that contains this `offset`. If no such block
+  // can be found, returns -1.
+  private def getBlockIndex(blockLocations: Array[BlockLocation], offset: Long): Int = {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala]: This method is only called once, should we inline it?

**Comment by cloud-fan** [comment]
Overall LGTM

File: sql/hive/src/test/scala/org/apache/spark/sql/sources/hadoopFsRelationSuites.scala (hunk 9 above)
Comment (liancheng) [inline sql/hive/src/test/scala/org/apache/spark/sql/sources/hadoopFsRelationSuites.scala]: Not necessarily, and probably only 1, since both files written are pretty small and will be merged into a single partition.

File: sql/hive/src/test/scala/org/apache/spark/sql/sources/hadoopFsRelationSuites.scala (hunk 8 above)
Comment (liancheng) [inline sql/hive/src/test/scala/org/apache/spark/sql/sources/hadoopFsRelationSuites.scala]: Yea, this is better, thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala (hunk 10 above)
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala]: This one is a little bit tricky, would like to keep it here for readability.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileScanRDD.scala [hunk 11]
```diff
@@ -86,4 +87,9 @@ class FileScanRDD(
   }
 
   override protected def getPartitions: Array[Partition] = filePartitions.toArray
+
+  override protected def getPreferredLocations(split: Partition): Seq[String] = {
+    val files = split.asInstanceOf[FilePartition].files
+    if (files.isEmpty) Seq.empty else files.maxBy(_.length).locations
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileScanRDD.scala]: Should we pick the top 3 host that hold most of the data?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala [hunk 12]
```diff
@@ -197,4 +204,26 @@ private[sql] object FileSourceStrategy extends Strategy with Logging {
...
+    blockLocations.indexWhere { b =>
+      b.getOffset <= offset && offset < b.getOffset + b.getLength
+    }
+  }
+
+  // Given locations of all blocks of a single file, `blockLocations`, and an `offset` position of
+  // the same file, find out the block that contains this `offset`, and returns location hosts of
+  // that block. If no such block can be found, returns an empty array.
+  private def getBlockHosts(blockLocations: Array[BlockLocation], offset: Long): Array[String] = {
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala]: Should we pick the hosts based on range that the host hold most of data in that range?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala (hunk 6 above)
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala]: We could do that in a separate PR (improve the locality with bounded complicity).

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileScanRDD.scala (hunk 11 above)
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileScanRDD.scala]: Updated. @cloud-fan also suggested the same strategy. Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala (hunk 12 above)
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala]: Makes sense, updated.

File: sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala [hunk 13]
```diff
@@ -624,20 +624,40 @@ class HDFSFileCatalog(
 
   def getStatus(path: Path): Array[FileStatus] = leafDirToChildrenFiles(path)
 
+  private implicit class LocatedFileStatusIterator(iterator: RemoteIterator[LocatedFileStatus])
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala]: Why do we need this?

File: sql/hive/src/test/scala/org/apache/spark/sql/sources/hadoopFsRelationSuites.scala [hunk 14]
```diff
@@ -668,6 +671,42 @@ abstract class HadoopFsRelationTest extends QueryTest with SQLTestUtils with Tes
       df.write.format(dataSourceName).partitionBy("c", "d", "e").saveAsTable("t")
     }
   }
+
+  test("Locality support for FileScanRDD") {
```
Comment (yhuai) [inline sql/hive/src/test/scala/org/apache/spark/sql/sources/hadoopFsRelationSuites.scala]: How about we add the following test cases?

* Every partition of the FileScanRDD has a single file.
* There are some large files. So, there are some partitions and every such a partition has a part of a file.
* If we are going to pick the top K locations that host the largest amount of data, let's also add tests.

File: sql/hive/src/test/scala/org/apache/spark/sql/sources/hadoopFsRelationSuites.scala (hunk 9 above)
Comment (yhuai) [inline sql/hive/src/test/scala/org/apache/spark/sql/sources/hadoopFsRelationSuites.scala]: If we expect them be packed in the same partition, let's add an `assert`.

File: sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala (hunk 13 above)
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala]: It's used [here][1]. `RemoteIterator` doesn't extend from `Iterator`. This implicit conversion helps to convert contents behind a `RemoteIterator` into an array.

[1]: https://github.com/apache/spark/pull/12153/files#diff-40c347747af9101e7e9fee52fc4120b8R656

**Comment by liancheng** [comment]
retest this please

**Comment by yhuai** [comment]
@liancheng Add more tests?

**Comment by liancheng** [comment]
retest this please

**Comment by liancheng** [comment]
Seems that Jenkins is down...

**Comment by yhuai** [comment]
test this please

File: sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala [hunk 15]
```diff
@@ -621,20 +621,40 @@ class HDFSFileCatalog(
...
+    extends Iterator[LocatedFileStatus] {
+
+    override def hasNext: Boolean = iterator.hasNext
+
+    override def next(): LocatedFileStatus = iterator.next()
+  }
+
   private def listLeafFiles(paths: Seq[Path]): mutable.LinkedHashSet[FileStatus] = {
     if (paths.length >= sqlContext.conf.parallelPartitionDiscoveryThreshold) {
       HadoopFsRelation.listLeafFilesInParallel(paths, hadoopConf, sqlContext.sparkContext)
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala]: Seems we also need to update the `listLeafFiles` that is called by `listLeafFilesInParallel`.

Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala]: Let's also have a test.

**Comment by yhuai** [comment]
Seems it is hanging in BucketedReadSuite again... Let's look at it together and figure out what's wrong.

**Comment by liancheng** [comment]
All the timeout in the Jenkins builds were due to a deadlock in `DAGScheduler`, and can be steadily reproduced locally by running the following test case

> BucketedReadSuite.only shuffle one side when 2 bucketed tables have different bucket keys.

This test case creates two bucketed tables both with 8 buckets and then joins them. Reducing 8 to 5 eliminates the deadlock. But I haven't figured out the real reason behind the deadlock. The deadlock also disappears if I remove FileScanRDD.preferredLocations(). Maybe that too many tasks are scheduled to the same place and exhausted some thread-pool?

**Comment by liancheng** [comment]
I'm closing this one for #12527, which is a rebased version of this one.
